state/keys_test.go

```diff
@@ -84,3 +86,71 @@ func TestMalformedKey(t *testing.T) {
 	keys := make(Keys)
 	require.False(keys.Add("", Read))
 }
+
+func TestHasPermissions(t *testing.T) {
+	allPerms := []Permissions{Read, Allocate, Write, None, All}
+
+	tests := []struct {
+		perm Permissions
+		has  []Permissions
+	}{
+		{
+			perm: Read,
+			has:  []Permissions{Read, None},
+		},
+		{
+			perm: Allocate,
+			has:  []Permissions{Read, Allocate, None},
+		},
+		{
+			perm: Write,
+			has:  []Permissions{Read, Write, None},
+		},
+		{
+			perm: None,
+			has:  []Permissions{None},
+		},
+		{
+			perm: All,
+			has:  allPerms,
+		},
+	}
+
+	for _, tt := range tests {
+		for _, perm := range allPerms {
+			permString := tt.perm.String()
+			ofPermString := perm.String()
+			has := tt.perm.Has(perm)
+
+			var name string
+			if has {
+				name = fmt.Sprintf("%s has %s", permString, ofPermString)
+			} else {
+				name = fmt.Sprintf("%s does not have %s", permString, ofPermString)
+			}
+
+			t.Run(name, func(t *testing.T) {
+				require := require.New(t)
+				contains := slices.Contains(tt.has, perm)
+				require.Equal(contains, has)
+			})
+		}
+	}
+}
+
+func (p Permissions) String() string {
```

Comment:
let's move this to `keys.go` instead of the test file

Reply:
Sure,
My reasoning with only having it in the test file is that these permissions can be added together and it would make this function very limiting in the amount of permissions it could represent. We may want to make it a bit more clever by making the string representation dynamic in the future

Reply:
I could do that as a followup

Reply:
I don't think this should ever be dynamic. I'd just add it into `keys.go`, this is fine to include outside of the test package imo and I don't think we should define a `String()` function on this type only within the test package.